### ISSUE TITLE
Fix incorrect handling of spilled registers in OOL code sections

### ISSUE THREAD

**Issue by fjeremic** [eclipse/omr#5977]
I’m working towards unifying the register allocator across codegens and I’m stumbling upon what seems to be a bug in the handling of spilled registers. I’ll use Power as an example here. The bug seems to be present in ARM, x86, and RISC-V as well.

In `doRegisterAssignment` we create and set the spilled register list which will hold all the spilled registers in the mainline instruction sequence. The point of this list is to know if a register was spilled in the hot path of an OOL code section. In such a scenario, this virtual register will be in a register on the entry to the OOL cold path and spilled on the exit. We need to know this while register allocating the OOL cold path because we must ensure that we properly spill the register on the exit merge branch. Failing to do this will result in an inconsistent register state at runtime (i.e. register is spilled on the hot path but not on the cold path, so future fills of this register could be loading garbage values).

It seems though according to the code that this spilled register list is initialized in `doRegisterAssignment` unconditionally!

https://github.com/eclipse/omr/blob/ffc8fc4a4a0e1490c4bcf70204ccef3d39ae7ef4/compiler/p/codegen/OMRCodeGenerator.cpp#L543-L555

However, when we register allocate the OOL code section we call `doRegisterAssignment` again, which means we will overwrite the original list, and the list will always be empty after allocating an OOL code section. This is problematic for spilled registers that live across an OOL code section.

I've included an picture of a whiteboard drawing which will aid the explanation:

![IMG_1246](https://user-images.githubusercontent.com/16259104/116754283-74a50780-a9d6-11eb-8895-a653327fe553.JPG)

Consider a backwards register allocation pass. Here is what happens on Power today:

1. At location A register X gets filled, meaning we generate a load instruction. The register is added to the SpilledRegisterList.
2. We register allocate OOL2 which will call `doRegisterAssignment` from within `TR_PPCOutOfLineCodeSection::assignRegisters`. This will reinitialize the SpilledRegisterList as per the above gist of code.
3. We finish allocating OOL2 and reach location B. At location B SpilledRegisterList does **not** contain register X.
4. We reach the branch to OOL1 in the mainline and we start allocating OOL1. Within `TR_PPCOutOfLineCodeSection::assignRegisters` we call `createCondForLiveAndSpilledGPRs` using our SpilledRegisterList. This happens here: https://github.com/eclipse/omr/blob/ffc8fc4a4a0e1490c4bcf70204ccef3d39ae7ef4/compiler/p/codegen/PPCOutOfLineCodeSection.cpp#L87-L102
Unfortunately SpilledRegisterList is empty because we lost track of register X from previous steps. This means we will **not** create a register dependency to ensure register X is in a spilled state on entry to the OOL cold path!

5. We register allocate OOL1. During the allocation we reach location C which is in the cold path of OOL1. At location C we want to make use of register X, so we generate a spill, which will be a store instruction. We proceed to use register X.
6. We reach the entry label of OOL1 which has all the dependencies created according to the register state from the mainline. Because of (4) we never generate a fill (a load) for register X.
7. We proceed back to register allocate the mainline, and reach location D, which makes use of register X. We generate a spill here (a store) since the backing storage of register X indicates that it is spilled.

The problem here is clear. Between locations D and C the value of register X is undefined because the instructions between locations D and C could have clobbered the register which contained the value at location D.

To fix this problem we would have had to properly load the value of register X at the entry point of OOL1, which we are currently not doing.

**Comment by fjeremic** [comment]
This is not a made up example. The following tree from OpenJ9 can cause the problem as observed on Z:

```
------------------------------
 n621n    (  0)  treetop                                                                              [0x50226FB1D0] bci=[-1,250,130] rc=0 vc=5094 vn=-
 n620n    ( 30)    anewarray  jitANewArray[#21  helper Method] [flags 0x400 0x0 ] (highWordZero Unsigned X!=0 allocationCanBeRemoved sharedMemory )  [0
 n618n    (  2)      iconst 14 (X!=0 X>=0 )                                                           [0x50226FB0E0] bci=[-1,248,130] rc=2 vc=5094 vn=-
 n619n    (  1)      loadaddr  java/lang/String[#382  Static] [flags 0x18307 0x0 ]                    [0x50226FB130] bci=[-1,250,130] rc=1 vc=5094 vn=-
------------------------------
enumReg = GPR_0715
classReg = GPR_0713 , resReg = &GPR_0717
acall: Internal Control Flow in OOL : false
------------------------------
 n621n    (  0)  treetop                                                                              [0x50226FB1D0] bci=[-1,250,130] rc=0 vc=5094 vn=-
 n620n    ( 29)    anewarray  jitANewArray[#21  helper Method] [flags 0x400 0x0 ] (in &GPR_0717) (highWordZero Unsigned X!=0 allocationCanBeRemoved sha
 n618n    (  1)      iconst 14 (in GPR_0714) (X!=0 X>=0 )                                             [0x50226FB0E0] bci=[-1,248,130] rc=1 vc=5094 vn=-
 n619n    (  0)      loadaddr  java/lang/String[#382  Static] [flags 0x18307 0x0 ] (in GPR_0713)      [0x50226FB130] bci=[-1,250,130] rc=0 vc=5094 vn=-
------------------------------

 [0x5065B70000]                        LGFI    GPR_0713,326918144
 [0x5065B70180]                        LHI     GPR_0714,0xe
 [0x5065B702D0]                        LGR     GPR_0715,GPR_0714
 [0x5065B70510]                        LGHI    GPR_0716,0x48
 [0x5065B706B0]                        AG      GPR_0716,#2628 96(GPR13)
 [0x5065B70840]                        CLG     GPR_0716,#2629 104(GPR13)
 [0x5065B73B20]                        Label L0155:     # (Start of internal control flow)
 [0x5065B70910]                        BRC     BL(0x2), Outlined Label L0150
 [0x5065B70AB0]                        LG      &GPR_0717,#2630 96(GPR13)
 [0x5065B70C40]                        STG     GPR_0716,#2631 96(GPR13)
 [0x5065B73680]                        IIHF    GPR_0715,326939648
 [0x5065B73850]                        STG     GPR_0715,#2632 0(&GPR_0717)
 [0x5065B739E0]                        PFD     2, #2633 256(&GPR_0717) # Prefetch for store
 [0x5065B741B0]                        ASSOCREGS {GPR0:D_GPR_0718:R}* {GPR1:D_GPR_0719:R}* {GPR2:GPR_0720:R} {GPR3:D_GPR_0721:R}* {GPR4:D_GPR_0722:R}*
  PRE:
 [0x5065B73C00]                        Label L0151:     # (End of internal control flow)
 POST:
 {AssignAny:GPR_0713:R} {AssignAny:GPR_0715:R} {AssignAny:&GPR_0717:R} {AssignAny:GPR_0716:R}

------------------------------

 [0x5065B70E90]                        Outlined Label L0150:     ; Denotes start of OOL for heap alloc
 [0x5065B71EB0]                        LGR     GPR_0749,GPR13
 [0x5065B72000]                        LGFR    GPR_0750,GPR_0714
 [0x5065B72AB0]                        ASSOCREGS {GPR0:D_GPR_0668:R}* {GPR1:D_GPR_0669:R}* {GPR2:D_GPR_0670:R}* {GPR3:D_GPR_0671:R}* {GPR4:D_GPR_0672:R
 [0x5065B733F0]                        ASSOCREGS {GPR0:D_GPR_0668:R}* {GPR1:GPR_0749:R} {GPR2:GPR_0750:R} {GPR3:GPR_0713:R} {GPR4:D_GPR_0672:R}* {GPR6:
  PRE:
 {GPR1:GPR_0749:R} {GPR2:GPR_0750:R} {GPR3:GPR_0713:R}
 [0x5065B724D0]                        BRASL   GPR_0751,0x0000000000000000
 POST:
 {GPR0:D_GPR_0718:D}* {GPR1:D_GPR_0719:D}* {GPR2:GPR_0720:D} {GPR3:D_GPR_0721:D}* {GPR4:D_GPR_0722:D}* {GPR6:D_GPR_0723:D}* {GPR7:D_GPR_0724:D}* {GPR14
 {FPR0:D_FPR_0725:D}* {FPR1:D_FPR_0726:D}* {FPR2:D_FPR_0727:D}* {FPR3:D_FPR_0728:D}* {FPR4:D_FPR_0729:D}* {FPR5:D_FPR_0730:D}* {FPR6:D_FPR_0731:D}* {FP
 {FPR8:D_FPR_0733:D}* {FPR9:D_FPR_0734:D}* {FPR10:D_FPR_0735:D}* {FPR11:D_FPR_0736:D}* {FPR12:D_FPR_0737:D}* {FPR13:D_FPR_0738:D}* {FPR14:D_FPR_0739:D}
 {VRF24:D_VRF_0741:D}* {VRF25:D_VRF_0742:D}* {VRF26:D_VRF_0743:D}* {VRF27:D_VRF_0744:D}* {VRF28:D_VRF_0745:D}* {VRF29:D_VRF_0746:D}* {VRF30:D_VRF_0747:
 [0x5065B734C0]                        LGR     &GPR_0717,GPR_0720
 [0x5065B735A0]                        BRC     NOP(0xf), Label L0151
```

In this example `GPR_0714` is commoned as per the reference count, and it gets used in both the mainline and OOL code sections. The other location that uses `GPR_0714` node `n618n` is another tree which looks identical, and there happens to be other OOL code sections in between, similar to the whiteboard example drawn previously.

**Comment by fjeremic** [comment]
@0xdaryl @gita-omr @knn-k @janvrany only Z and AARch64 seem to be doing the correct thing here. I'm going to fix this in #5975 which is where I first noticed the problem by removing the following `!self()->isOutOfLineColdPath()` check from `doRegisterAssignment` on Z:

https://github.com/eclipse/omr/blob/ffc8fc4a4a0e1490c4bcf70204ccef3d39ae7ef4/compiler/z/codegen/OMRCodeGenerator.cpp#L1718-L1722

Here is how it is on Power for comparison:

https://github.com/eclipse/omr/blob/ffc8fc4a4a0e1490c4bcf70204ccef3d39ae7ef4/compiler/p/codegen/OMRCodeGenerator.cpp#L554-L555

Just letting you know this is happening.
